### PR TITLE
(WIP) Guidance Refactor

### PR DESCRIPTION
**Pull request by LorenLuke**
EARLY Work in Progress

**When merged this pull request will:**
- Replace ace_missileguidance in whole, while incorporating existing work from previous framework and their contributors into the project
- Create more streamlined, documented, and mod-friendly framework for ace guided projectile module
- Enable guidance capability for additional munitions, including bombs, mortars, artillery shells, rifle grenades, etc.
- Modularise existing sensor types for usage as guidance sensors, and incorporate currently missing or unincluded control types, including GPS, Semi-Active Laser, Active Radar, Passive Radar, Manual-Command LOS, Semi-Active Command LOS, Predicted LOS, Infrared, and Electro-Optical Image Comparison.
- Modularise and condense existing flight profiles into several types to streamline config entries and simplify compat creation for 3rd party mods, and add additional flight modes not included, such as Surface-to-Surface Ballistic Missile profiles, direct and indirect fire ballistic profiles, etc.
- Enable multiple fire modes for given projectiles, enabling players to easily switch seekers or attack profiles within the vanilla game's framework
- Integrate weapons platforms' capabilities for improved user experience on using otherwise cumbersome or clunky systems to perform guidance tasks, e.g. allowing players to use a pilot camera for selection of targets for GPS-Guided Munitions.

- [ ] Functionality Pass
    - [ ] PreGuidance (optional)
    - [x] PFH
    - [ ] Seekers
        - [x] IR
        - [x] GPS/INS
        - [x] ElectroOptical (Imaging Infrared)
        - [ ] SACLOS
        - [x] MCLOS
        - [x] Predicted LOS
        - [x] Semi-Active Laser
        - [ ] Active Radar Homing
        - [x] Passive Radar Homing
    - [ ] Attack Profiles 
        \- [ ] ~~FIM~~
        \- [ ] ~~FGM~~
            \- [ ] ~~Direct~~
            \- [ ] ~~Top-Attack~~
            \- [ ] ~~Overfly~~
        \- [ ] ~~ATGM~~
            \- [ ] ~~Direct~~
            \- [ ] ~~Top-Attack~~
            \- [ ] ~~Overfly~~
        \- [ ] ~~SAM~~
        \- [ ] ~~AAM~~
        - [ ] SSM (merged FGM, AGM, ATGM)
            - [x] Direct (/LOBL)
            - [ ] Top-Attack (/LOAL-Hi)
            - [ ] Overfly (/LOAL-Lo)
            - [ ] Terrain Following 

        - [ ] AIM (merged FIM, AAM, SAM)
        - [x] SSBM
        - [x] GBU
        - [x] Indirect fire (shell)
        - [x] Direct fire (shell)

    - [x] Fuzing
        - [x] Proximity
        - [x] Altitude
        - [x] Range
        - [x] Time
        - [x] Point in space

- [ ] Debug Pass
    - [ ] Privatize variables
    - [ ] Include Traces
    - [ ] Include Debug visualizations

- [ ] Systems Integration Pass
    - [ ] Remove DEBUG ammo classes
    - [ ] Integrate existing ACE Module frameworks
        - [ ] Expand firemodes for existing vanilla and ACE weapons for better UX
    - [ ] Allow ace_missileguidance backwards compatibility

- [ ] Documentation Pass
    - [ ] Comment pass all scripts
    - [ ] Document all modes of usage for each weapon

- [ ] Optimization Pass
    - [ ] Review arguments
    - [ ] Pass optimization review

- [ ] Documentation Cleanup Pass
    - [ ] Review documentation

- [ ] Review Ready

___________

Stretch goals:

- [ ]  In-flight integration of MicroDAGR with aircraft pilot cameras, and GPS guided weapons, including 'cycle points on weapon release'.

- [ ] Multiple seekers on the same projectile

Inclusion of additional weapons systems/capabilities for use with guidance-
- [ ] TITAN AP/AT as a Spike television guided projectile
- [ ] Fuzed MAAWS HE rounds
- [ ] Ballistic missile for MK41 VLS
- [ ] Fuzed HEAT-MP shells
- [ ] 82mm laser/infrared/GPS guided rounds 
- [ ] 155mm GPS guided rounds

### CODE REVIEW

**Comment by Drofseh** [comment]
Won't changing addon name break functionality for every mod that has adopted the current classes? That seems like a bad idea to me.

Even if the system is being expanded to included guidance for other weapons, such as bombs or grenades, those are also missiles in the strictest sense of the word.

https://github.com/LorenLuke/ACE3/blob/guidance/addons/guidance/CfgAmmo.hpp#L12
https://github.com/LorenLuke/ACE3/blob/guidance/addons/guidance/functions/fnc_extractInfo.sqf#L32-L52

File: addons/guidance/functions/fnc_extractInfo.sqf [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        if (getPilotCameraTarget _shooter select 0) then {
+            _targetPos = getPilotCameraTarget _shooter select 1;
+            _targetVector = (ATLtoASL (_shooter modeltoWorld (getPilotCameraPosition _shooter))) vectorFromTo _targetPos;
+            if(getPilotCameraTarget _shooter select 2 != objNull) then {
+                _target = getPilotCameraTarget _shooter select 2;
+            };
+    //if game-locked target
+        } else {
+            if (!(isNull cursorTarget)) then {
+                _target = cursorTarget;
```
Comment (TheCandianVendingMachine) [inline addons/guidance/functions/fnc_extractInfo.sqf]: `missileTarget _projectile` will give a more accurate game locked target

**Comment by LorenLuke** [comment]
@Drofseh 
I would argue that it depends how it's done. For example, if I were to just comb through the scripts (not even the configs, but the scripts) and put a provision looking for the exact same thing with the additional search of 'ace_missileguidance', then nothing is lost. The idea of changing the name to a new system enables a search for the legacy system and its respective variables (which may or may not be obsolete in the new system) and converting them into something usable under the new guidance framework. By having the legacy system in place, it'll enable usage and of the new functionalities afforded by the new guidance module, and still be backwards compatible with old values and compats.

**Comment by Mike-MF** [comment]
Closing due to age and/or inactivity, unlikely to be finished or has to be reworked due to other changes over the years. May be re-opened if someone finds it useful, even better just open a new up-to-date PR.

**Pull request closed without merging**